pkg/kubelet/cm/dra/plugin/client_test.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+	drapbv1 "k8s.io/kubelet/pkg/apis/dra/v1alpha2"
+)
+
+type fakeGRPCServer struct {
+	drapbv1.UnimplementedNodeServer
+}
+
+func (f *fakeGRPCServer) NodePrepareResource(ctx context.Context, in *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+	return &drapbv1.NodePrepareResourceResponse{CdiDevices: []string{"dummy"}}, nil
+}
+
+func (f *fakeGRPCServer) NodeUnprepareResource(ctx context.Context, in *drapbv1.NodeUnprepareResourceRequest) (*drapbv1.NodeUnprepareResourceResponse, error) {
+	return &drapbv1.NodeUnprepareResourceResponse{}, nil
+}
+
+type tearDown func()
+
+func setupFakeGRPCServer(t *testing.T) (string, tearDown) {
+	p, err := os.MkdirTemp("", "dra_plugin")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	closeCh := make(chan struct{})
+	addr := filepath.Join(p, "server.sock")
+	teardown := func() {
+		close(closeCh)
+		os.RemoveAll(addr)
+	}
+
+	listener, err := net.Listen("unix", addr)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	s := grpc.NewServer()
+	fakeGRPCServer := &fakeGRPCServer{}
+	drapbv1.RegisterNodeServer(s, fakeGRPCServer)
+
+	go func() {
+		go s.Serve(listener)
+		<-closeCh
+		s.GracefulStop()
+	}()
+
+	return addr, teardown
+}
+
+func TestGRPCConnIsReused(t *testing.T) {
+	addr, teardown := setupFakeGRPCServer(t)
+	witness := make(map[string]int)
+	wg := sync.WaitGroup{}
+	m := sync.Mutex{}
+
+	plugin := &Plugin{
+		endpoint: addr,
+	}
+
+	if err := plugin.CreateGRPCConn(); err != nil {
+		t.Fatal(err)
+	}
+
+	ctime := time.After(100 * time.Millisecond)
+	// wait a bit for the connection to be in ready state before performing our test
+	for !plugin.IsConnReady() {
+		select {
+		case <-ctime:
+			t.Fatal("gRPC connection was not ready")
+		default:
+		}
+		time.Sleep(time.Millisecond)
+	}
```

Comment:
What is the potential for flakiness here? Is there precedent for putting arbitrary timeouts like this when testing grpcConnections?

Reply:
I don't know if there is a precedent but I felt like it was necessary to test the conn was ready before starting my test.
I put a timeout to avoid the test to hand indefinitely. 
If you don't feel comfortable I will remove that part.
Regarding the potential of flakiness, below the output of stress testing this specific test
<img width="372" alt="Screenshot 2023-07-09 at 17 56 59" src="https://github.com/kubernetes/kubernetes/assets/9576370/3ff02f0a-f114-4c32-9a5a-2aec40d9169e">